cluster-autoscaler/proposals/resourceslack.md

```diff
@@ -0,0 +1,266 @@
+This behavior of scaling up implies that, in a worst case, an unschedulable pod must wait until an eligible node is created, started, and then gets ready before it is finally scheduled.
+A node takes a much longer time than that of a pod to start and become ready and therefore it results in pods to occasionally delayed for several minutes to be horizontally scaled out.
+A possible work-around can be done today is to lower the threshold of HPA to trigger a scaling up of a replicaset/deployment earlier but it reduces total resource utilization when there are many replicasets and deployments required the work-around.
+Having the resource slack feature allows us to scale out pods faster while minimizing impact to resource utilization.
+
+## Design
+
+To achieve Resource Slack, introduce placeholder pods that will live only inside of ClusterAutoscaler.
+They are created to virtually request certain amounts of CPU and memory so that they influence all of the scale-up/scale-down decision made in CA, but won't be present in the scheduler running inside an actual K8S cluster, so the new pods will be able to pick the place of the placeholders and push the placeholders elsewhere (and make them trigger the scale-up/down).
+
+### Requirements
+
+Version 1
+
+* [R1] Works on relatively large clusters today's CA used to work
+  * dozens of nodes, hundreds of pods
+  * In general we don't want to have O(pods * nodes) operations every single loop
+  * [R1-1] On large cluster there is a huge number of pods to be scheduled. And the pods are mostly the same. If a pod from one replica set X failed to schedule on node A then there is no point to in trying to fit the next pod from X on A. Thus there should be just a single loop over all nodes (ref https://github.com/kubernetes/autoscaler/pull/56#discussion_r115942312)
+  * [R1-2] The scale down performance strongly depends that the pods don't change every iteration. So if you placed placeholder-X-123 on node A then in the next iteration it should also land on A or the scale down performance will be heavily impacted (ref https://github.com/kubernetes/autoscaler/pull/56#discussion_r115942312)
+* [R2] Keep the specified amount of extra capacity at minimum
+  * If we allowed users to annotate their k8s pods so that they are mirrored to equivalent placeholder pods like https://github.com/kubernetes/autoscaler/pull/77#discussion_r117084488,
+    CA may actually produce much more extra capacity than specified by the user.
+
+    Let's say an user specified CA to reserve 10% extra capacity, but also marked a pod which requests cpu and memory more than 10% of current capacity. CA will result in creating a placeholder pod which consumes resource more than 10%.
+    So, the "10%" in this case is just a minimum requirement. CA may reserve more if the user instructed CA to do so by annotations. That's why I wrote `at minimum`. It is impossible to specify CA an exact rate of extra capacity when we want to suppoprt [R6] and [R7].
```

Comment:
Wait, so if I had a deployment with 10 replicas it would result in slack for additional 10 pods? Doubling the size of every deployment, rc, etc. feels very excessive.

Also would the extra capacity coming from annotated pods count toward the general number ("10%")? Or would that be in addition?